runtime/include/tt/runtime/detail/logger.h

```diff
@@ -0,0 +1,378 @@
+  log_error_(LogAlways, args...);
+}
+
+template <typename... Args>
+static void log_fatal_(LogType type, Args const &...args) {
+  Logger::get().log_level_type(Logger::Level::Fatal, type, args...);
+}
+
+template <typename... Args> static void log_fatal_(Args const &...args) {
+  log_fatal_(LogAlways, args...);
+}
+
+template <typename... Args>
+[[noreturn]] static void
+tt_throw_(char const *file, int line, char const *assert_type,
+          char const *condition_str, Args const &...args) {
+  std::stringstream trace_message_ss = {};
+  trace_message_ss << "\n";
+  trace_message_ss << bold << assert_type << " @ " << file << ":" << line
+                   << ": " << condition_str << std::endl;
+  trace_message_ss << "backtrace:\n";
+  trace_message_ss << detail::backtrace_to_string(100, 3, " --- ");
+  trace_message_ss << reset_text_attrs << std::flush;
+  log_fatal_(args..., trace_message_ss.str());
+  Logger::get().flush();
+  throw std::runtime_error("Fatal error");
+}
+
+#undef LOGGER_TYPES
+
+} // namespace tt::runtime::logger
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wgnu-zero-variadic-macro-arguments"
+#ifdef TT_RUNTIME_DEBUG
+#define LOG_TRACE(log_type, ...)                                               \
+  do {                                                                         \
+    if (::tt::runtime::logger::Logger::get().log_level_enabled(                \
+            ::tt::runtime::logger::Logger::Level::Trace) &&                    \
+        ::tt::runtime::logger::Logger::get().log_type_enabled(log_type))       \
+      ::tt::runtime::logger::log_trace_(                                       \
+          log_type,                                                            \
+          ::tt::runtime::logger::bold + std::string(__FILE__) + ":" +          \
+              std::to_string(__LINE__) +                                       \
+              ::tt::runtime::logger::reset_text_attrs,                         \
+          ##__VA_ARGS__);                                                      \
+  } while (0)
+
+#define LOG_DEBUG(...)                                                         \
+  do {                                                                         \
+    if (::tt::runtime::logger::Logger::get().log_level_enabled(                \
+            ::tt::runtime::logger::Logger::Level::Debug)) {                    \
+      ::tt::runtime::logger::log_debug_(__VA_ARGS__);                          \
+    }                                                                          \
+  } while (0)
+#else
+#define LOG_TRACE(...) ((void)0)
+#define LOG_DEBUG(...) ((void)0)
+#endif
+
+#define LOG_INFO(...)                                                          \
+  do {                                                                         \
+    if (::tt::runtime::logger::Logger::get().log_level_enabled(                \
```

Comment:
Why are we wrapping the calls with checks if log level/types are enabled? The `log_level_type()` is already doing that.

Reply:
So previously on bbe we hit issue with logger slowing us down, because if I did something like `LOG_INFO("My value is {}", some_function())` and `some_function()` has non-trivial overhead, we would get a perf hit. Checking this prior to log_level_type ensures that the function will not get called if we don't want to log this particular level/type. This is also why we defined all these logger APIs as macros, instead of template functions directly.

Reply:
Makes sense, thanks!